Issue/16710 - Implement logic for referrer details scene

nikola-milicevic opened the pull request:
Part of [#16710](https://github.com/wordpress-mobile/WordPress-iOS/issues/16710)
Depends on [#16715](https://github.com/wordpress-mobile/WordPress-iOS/pull/16715)

This PR adds logic for marking / unmarking referrer as spam.

### To test:

1. Tap on 'Stats' button located in 'My Site' tab.
2. Scroll down to referrers section.
3. Tapping on a referrer displays referrer details scene.
4. Tapping on 'View more', and then on a referrer also displays referrer details scene.
5. Tap on 'Mark as spam' or 'Mark as not spam'

### Outcome:


https://user-images.githubusercontent.com/12729242/122833573-7e1f6180-d2ed-11eb-912a-962b50525507.mp4



### Regression Notes
1. Potential unintended areas of impact
n/a

2. What I did to test those areas of impact (or what existing automated tests I relied on)
n/a

3. What automated tests I added (or what prevented me from doing so)
n/a

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

nikola-milicevic committed:
Inject data into referrer details scene.

# Conflicts:
#	WordPress/Classes/ViewRelated/Stats/Referrer Details/ReferrerDetailsViewModel.swift

nikola-milicevic committed:
Build details rows, show real data inside details scene

nikola-milicevic committed:
Add row action.

nikola-milicevic committed:
Add ReferrerDetailsViewModelDelegate protocol.

nikola-milicevic committed:
Implement displayWebViewWithURL.

nikola-milicevic committed:
Implement toggleSpamState delegate method, update referrer in StatsPeriodStore

nikola-milicevic committed:
Remove unused methods.

nikola-milicevic committed:
Refresh table view on successful toggling of isSpam referrer state.

emilylaguna reviewed (changes requested):
Heya, nice work. The new design looks good! I noticed a few things when testing:

- When tapping on the <kbd>Mark as Spam</kbd> button I found it a bit odd to then have to tap another <kbd>Mark as Spam</kbd> button the action sheet to trigger the action. What do you think about removing the extra step?
- When testing with slower internet it can feel as though nothing is happening after I choose to mark an item as spam. I wonder if we should display an indicator?
- What do you think about showing the WebView instead of the details view if a user taps on an item that can not be marked as spam and has 1 row to display? Example being: `WordPress.com Reader`
- The app crashes when tapping <kbd>Mark as Spam</kbd> table cell on iPad

nikola-milicevic committed:
Remove action sheet.

nikola-milicevic committed:
Display webview when tapped on a referrer with no children which cannot be marked as spam

nikola-milicevic committed:
Add loader to spam action cell.

nikola-milicevic committed:
Configure spam action cell with isLoading value.

nikola-milicevic committed:
Set loading state to appropriate value.

nikola-milicevic commented:
> * When tapping on the Mark as Spam button I found it a bit odd to then have to tap another Mark as Spam button the action sheet to trigger the action. What do you think about removing the extra step?

Yeah, totally agree.

> * When testing with slower internet it can feel as though nothing is happening after I choose to mark an item as spam. I wonder if we should display an indicator?

I was thinking about that, but didn't implement it since I wasn't sure about the design. As discussed over slack, now the action text is replaced with the loader while loading is in progress. I updated the video in the description.

> * What do you think about showing the WebView instead of the details view if a user taps on an item that can not be marked as spam and has 1 row to display? Example being: `WordPress.com Reader`

Yes, definitely feels better.

> * The app crashes when tapping Mark as Spam table cell on iPad

Although these changes are not tested on iPad yet (will be done in future PR), the crash shouldn't happen now since we got rid of action sheet 😊

Thanks for the comments, I updated the code!

nikola-milicevic committed:
Add dark mode support.

nikola-milicevic committed:
Use safe area anchors for correct landscape layout.

emilylaguna commented on WordPress/Classes/ViewRelated/Stats/Referrer Details/ReferrerDetailsViewModel.swift:
```diff
@@ -20,6 +21,10 @@ extension ReferrerDetailsViewModel {
         self.data = data
     }
+
+    func setLoadingState(_ value: Bool) {
```
Question: Is there a benefit to exposing just the setter vs making the entire var public?

nikola-milicevic replied:
I think it's easier to track what's going on that way, it's more explicit. But really I don't have any strong argument against the other approach. Should I leave it as is, or make the property non-private?

emilylaguna replied:
Ah makes sense. I think we can leave it the way it is right now. I don't have a strong opinion either way.

emilylaguna commented on WordPress/Classes/ViewRelated/Stats/Referrer Details/ReferrerDetailsCell.swift:
```diff
@@ -29,6 +29,7 @@ extension ReferrerDetailsCell {
 private extension ReferrerDetailsCell {
     func setupViews() {
         selectionStyle = .none
+        backgroundColor = .listForeground
```
What do you think about setting this in WPStyleGuide?

emilylaguna reviewed (changes requested):
Nice, I think the flow makes more sense now and the loading spinner looks good! 👍 

I left a comment and question for you.

nikola-milicevic committed:
Use WPStyleGuide for cell background color.

emilylaguna reviewed (approved):
LGTM!
